appcelerator/titanium_mobile
# fix(android): Modified "Ti.Database" install() and open() methods

**JIRA:**
- https://jira.appcelerator.org/browse/TIMOB-26878
- https://jira.appcelerator.org/browse/TIMOB-26877

**Summary:**
- [TIMOB-26878](https://jira.appcelerator.org/browse/TIMOB-26878) Fixed bug where `Ti.Database.install()` would fail if destination directory did not exist.
  * Fixed by creating directory tree via `mkdirs()` before copying source file.
  * Note that this issue commonly happens on Android 6.0.
- [TIMOB-26877](https://jira.appcelerator.org/browse/TIMOB-26877) Modified `Ti.Database.install()` to throw an exception if given source database file was not found.
  * Old behavior would create a new database file instead, which was causing confusion. If you want to create a new/clean database, then use the `open()` method instead.
  * Now matches iOS' behavior.
- Modified `Ti.Database.open()` to check if SQLite returned `null` or not.
  * Old code was assuming it was not `null` and attempting to log the database's name, causing an exception.
- Removed logging of exception before re-throwing it in `install()` and `open()` methods.
  * Not needed since Titanium's unhandled exception handler will log it anyways. Prevents the exception from being logged twice.

**Test:**
1. Create a Classic Titanium app.
2. Set the "tiapp.xml" property "analytics" to `false` as shown below.
3. Set up the "app.js" with the below code.
4. Copy the `Database1.sqlite` file from [TIMOB-26878](https://jira.appcelerator.org/browse/TIMOB-26878) to your project's "Resources" directory.
5. Build and run on Android 6.0.
6. Verify that the following was logged on startup.
```
[INFO] :   - Key: 'CompanyName', Value: 'Axway'
[INFO] :   - Key: 'TestMessage', Value: 'Hello World'
```
7. Back out of the app.
8. Restart the app.
9. Verify that `- Key: 'LastLaunchTime'...` was logged. 

---
**tiapp.xml**
```xml
<?xml version="1.0" encoding="UTF-8"?>
<ti:app xmlns:ti="http://ti.appcelerator.org">
	<analytics>false</analytics>
</ti:app>
```

---
**app.js**
```javascript
var dbConnection = Ti.Database.install("Database1.sqlite", "Database1");
var resultSet = dbConnection.execute("SELECT key, value FROM properties");
Ti.API.info("@@@ Database Table 'properties' row count: " + (resultSet ? resultSet.rowCount : "<null>"));
if (resultSet) {
	for (; resultSet.isValidRow(); resultSet.next()) {
		Ti.API.info("- Key: '" + resultSet.field(0) + "', Value: '" + resultSet.field(1) + "'");
	}
}
dbConnection.execute("INSERT OR REPLACE INTO properties(key, value) VALUES (?, ?)", "LastLaunchTime", new Date());
dbConnection.close();
```
